ankimorphs/text_highlighting.py

```diff
@@ -1,238 +1,386 @@
-        else:
-            morph_status = "known"
+    def __init__(self, start: int, end: int, status: str):
+        super().__init__(start, end)
+        self.status = status
+
+    def open_len(self) -> int:
+        """Len of the open tag, useful for setting string splice offsets."""
+        return len(self.open())
+
+    def open(self) -> str:
+        return f'<span morph-status="{self.status}">'
+
+    def close(self) -> str:
+        return "</span>"
 
-        # escaping special regex characters is crucial because morphs from malformed text
-        # sometimes can include them, e.g. "?몇"
-        regex_pattern: str = f"{re.escape(morph.inflection)}"
-        morph_matches = re.finditer(
-            regex_pattern, text_to_highlight, flags=re.IGNORECASE
+    def inject(self, target: str) -> str:
+        """Put this morph into the given string."""
+
+        return (
+            target[: self.start]
+            + self.open()
+            + target[self.start : self.end]
+            + self.close()
+            + target[self.end :]
         )
 
-        for morph_match in morph_matches:
-            start_index = morph_match.start()
-            end_index = morph_match.end()
-            morph_len = end_index - start_index
+    def __repr__(self) -> str:
+        return f"Range: {self.start}-{self.end}. Status: {self.status}."
+
+
+class TextHighlighter:
+    """Represents an expression to highlight. Tracks 2 sets of data, one for rubies the other
+    for morph status. all the magic happens in _process() where we merge them together on top
+    of the base string."""
+
+    def __init__(self, expression: str, morphs_and_statuses: list[MorphAndStatus]):
+        self._highlighted: str | None = None
+        self.expression: str = expression
+        self.rubies: deque[RubyRange] = deque()
+        self.statuses: deque[StatusRange] = deque()
+
+        self._tag_rubies()
+        self._tag_morphemes(self.expression.lower(), morphs_and_statuses)
 
-            # the morph_match.group() maintains the original letter casing of the
-            # morph found in the text, which is crucial because we want everything
-            # to be identical to the original text.
-            span_elements.append(
-                SpanElement(morph_match.group(), morph_status, start_index, end_index)
+    def _tag_rubies(self) -> None:
+        """Populate internal deque of found ruby locations."""
+
+        end = 0
+
+        while True:
+            match = text_preprocessing.ruby_regex.search(self.expression, pos=end)
+
+            if not match:
+                break
+
+            end = match.start() + len(match.group(1))
+
+            self.rubies.append(
+                RubyRange(match.start(), end, match.group(1), match.group(2))
+            )
+            self.expression = (
+                self.expression[: match.start()]
+                + match.group(1)
+                + self.expression[match.end() :]
```

Comment:
This block now tracks the prior match, and ensures that the current search does not match before that position. Nice catch on that bug!!